alacritty/src/input.rs

```diff
@@ -984,18 +984,28 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
         let mode = BindingMode::new(self.ctx.terminal().mode(), self.ctx.search_active());
         let mouse_mode = self.ctx.mouse_mode();
         let mods = self.ctx.modifiers().state();
+        let mut actions = vec![];
 
-        for i in 0..self.ctx.config().mouse_bindings().len() {
-            let mut binding = self.ctx.config().mouse_bindings()[i].clone();
-
-            // Require shift for all modifiers when mouse mode is active.
-            if mouse_mode {
-                binding.mods |= ModifiersState::SHIFT;
-            }
+        // If mouse mode is active, also look for bindings without Shift
+        let fallback_mods = mods & !ModifiersState::SHIFT;
+        let mut use_fallback = mouse_mode && mods.contains(ModifiersState::SHIFT);
 
+        for binding in self.ctx.config().mouse_bindings() {
             if binding.is_triggered_by(mode, mods, &button) {
-                binding.action.execute(&mut self.ctx);
+                if use_fallback {
+                    // Throw away fallback actions, stop looking for them
+                    actions.clear();
+                    use_fallback = false;
+                }
+                actions.push(binding.action.clone());
             }
+            if use_fallback && binding.is_triggered_by(mode, fallback_mods, &button) {
+                actions.push(binding.action.clone());
+            }
+        }
+
+        for action in actions {
+            action.execute(&mut self.ctx);
```

Comment:
Could you simply take the code which was there before and do 2 loops? The amount of bindings is very low so it's better than `vec` approach.

Just do normal loop without `mouse_mode` part and set some flag to `true` when you actually triggered some binding.

And then, `if mouse_mode && !sent_binding` do the ho the same loop, but with `SHIFT` applied? First loop won't need to `clone` even, I think.